storage/src/key_value/agent/block_commit.rs

```diff
@@ -0,0 +1,291 @@
+        let mut side_chain_path = VecDeque::new();
+        let header = self.get_block_header(hash)?;
+        let canon_height = self.canon_height()?;
+        let mut parent_hash = header.previous_block_hash;
+        for i in 0..=oldest_fork_threshold {
+            // check if the part is part of the canon chain
+            match self.get_block_state(&parent_hash)? {
+                // This is a canon parent
+                BlockStatus::Committed(block_num) => {
+                    // Add the children from the latest block
+                    if block_num + oldest_fork_threshold - i < canon_height as usize {
+                        debug!("exceeded maximum fork length in extended path");
+                        return Ok(ForkDescription::TooLong);
+                    }
+                    let longest_path = self.longest_child_path(hash)?;
+                    side_chain_path.extend(longest_path);
+
+                    return Ok(ForkDescription::Path(ForkPath {
+                        base_index: block_num as u32,
+                        path: side_chain_path.into(),
+                    }));
+                }
+                // Add to the side_chain_path
+                BlockStatus::Uncommitted => {
+                    side_chain_path.push_front(parent_hash.clone());
+                    parent_hash = self.get_block_header(&parent_hash)?.previous_block_hash;
+                }
+                BlockStatus::Unknown => {
+                    return Ok(ForkDescription::Orphan);
+                }
+            }
+        }
+        Ok(ForkDescription::TooLong)
+    }
+
+    pub(super) fn commit_transaction(
+        &mut self,
+        sn_index: &mut u32,
+        cm_index: &mut u32,
+        memo_index: &mut u32,
+        transaction: &SerialTransaction,
+    ) -> Result<Vec<(Digest, u32)>> {
+        let mut commitments = vec![];
+        // we are leaving validation to the ledger
+        for serial in transaction.old_serial_numbers.iter() {
+            self.inner.store(
+                KeyValueColumn::SerialNumber,
+                &serial[..],
+                &(*sn_index).to_le_bytes()[..],
+            )?;
+            *sn_index += 1;
+        }
+
+        for commitment in transaction.new_commitments.iter() {
+            self.inner.store(
+                KeyValueColumn::Commitment,
+                &commitment[..],
+                &(*cm_index).to_le_bytes()[..],
+            )?;
+            commitments.push((commitment.clone(), *cm_index));
+            *cm_index += 1;
+        }
+
+        self.inner.store(
+            KeyValueColumn::Memo,
+            &transaction.memorandum[..],
+            &(*memo_index).to_le_bytes()[..],
+        )?;
+        *memo_index += 1;
+
+        Ok(commitments)
+    }
+
+    pub(super) fn commit_block(&mut self, block_hash: &Digest, ledger_digest: &Digest) -> Result<BlockStatus> {
+        let canon = self.canon()?;
+        let block = self.get_block(block_hash)?;
```

Comment:
`get_block` should happen only after the `get_block_state` check, in order to speed things up for the already-`Committed` blocks